### ISSUE TITLE
submodules break release-opinion

### ISSUE THREAD

**Issue by paleozogt** [ajoberstar/grgit#60]
Using submodules breaks grgit's release-opinion.  Even when run directly on a tag (e.g., '1.1.0') it will produce versions as if it had uncommitted changes (e.g., '1.1.1-dev.0.uncommitted+2db5eac').

I've written up a minimal project that reproduces this problem [here](https://github.com/paleozogt/paleozogt-sandbox).

If we checkout the project:

```
> git --version
git version 2.4.0
> 
> git clone git@github.com:paleozogt/paleozogt-sandbox.git
   ...snip...
> 
> cd paleozogt-sandbox 
```

and then checkout a tag that doesn't use a submodule, it works fine:

```
> git checkout 1.0.0
   ...snip...
> git describe
1.0.0
>
> ./gradlew tasks
project version 1.0.0
git describe 1.0.0
:tasks
   ...snip...
```

But if we then checkout a tag that does use a submodule, it produces 'dirty' version numbers (despite git describe being simple):

```
> git checkout 1.1.0
   ...snip...
>
> git submodule sync
> git submodule update --init
   ...snip...
>
> git describe
1.1.0
>
> ./gradlew tasks
Inferred project: paleozogt-sandbox, version: 1.1.1-dev.0.uncommitted+2db5eac
project version 1.1.1-dev.0.uncommitted+2db5eac
git describe 1.1.0
:tasks
   ...snip...
```

One other thing that popped out at me is that in the submodule-using tag, grgit outputs 'Inferred project ...', but it doesn't output this in the tag that doesn't use a submodule.  This seems to imply that grgit is going down different code-paths for the two tags.

**Comment by paleozogt** [comment]
fyi I also tried this using git 1.9.5 and got the same results.

**Comment by ajoberstar** [comment]
Adding `println git.status()` shows the following in your `1.1.0` tag.

```
org.ajoberstar.grgit.Status(staged:org.ajoberstar.grgit.Status$Changes(added:[], modified:[], removed:[]), unstaged:org.ajoberstar.grgit.Status$Changes(added:[], modified:[build.gradle], removed:[grgit]))
```

This is just showing that the root cause is that Grgit interprets the `grgit` entry as a "removed" but "unstaged" change. Looking at where this comes from, JGit is interpreting that submodule as a "missing" file.

Still digging to see why that's the case.

**Comment by ajoberstar** [comment]
OK, next I tried:

```
git submodule init
git submodule update
```

Now rerunning `gradlew tasks` shows:

```
project version: 1.1.0
git describe: 1.1.0
```

So as long as the submodules are fully checked out, it looks like it works just fine. Let me know if that also resolves it for you.

**Comment by ajoberstar** [comment]
Feel free to reopen if that didn't resolve the issue for you.

**Comment by paleozogt** [comment]
Sorry I wasn't able to get back to you promptly.  I'm not getting the behavior you're describing-- the git submodule init/update doesn't change how grgit/jgit sees the submodule:

```
> git --version
git version 2.4.0
>
> git clone git@github.com:paleozogt/paleozogt-sandbox.git
Cloning into 'paleozogt-sandbox'...
remote: Counting objects: 411, done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 411 (delta 4), reused 0 (delta 0), pack-reused 399
Receiving objects: 100% (411/411), 1.32 MiB | 1.42 MiB/s, done.
Resolving deltas: 100% (76/76), done.
Checking connectivity... done.
>
> cd paleozogt-sandbox 
>
> git describe
1.2.0
>
> git submodule init
Submodule 'grgit' (git@github.com:ajoberstar/grgit.git) registered for path 'grgit'
>
> git submodule update
Cloning into 'grgit'...
remote: Counting objects: 4727, done.
remote: Compressing objects: 100% (238/238), done.
remote: Total 4727 (delta 171), reused 0 (delta 0), pack-reused 4452
Receiving objects: 100% (4727/4727), 1.72 MiB | 1.24 MiB/s, done.
Resolving deltas: 100% (2989/2989), done.
Checking connectivity... done.
Submodule path 'grgit': checked out '0da1151ce5ad67ab251ea5cd080f3531e8cad3c0'
>
> git status
On branch grgit_tests
Your branch is up-to-date with 'origin/grgit_tests'.
nothing to commit, working directory clean
>
> ./gradlew tasks
Inferred project: paleozogt-sandbox, version: 1.2.1-dev.0.uncommitted+1f285fc
project version 1.2.1-dev.0.uncommitted+1f285fc
git describe 1.2.0
git status org.ajoberstar.grgit.Status(staged:org.ajoberstar.grgit.Status$Changes(added:[], modified:[], removed:[]), unstaged:org.ajoberstar.grgit.Status$Changes(added:[], modified:[grgit], removed:[]))
   ...snip...
> 
``` 

I added the print of `git.status()`-- like you said, it looks like JGit thinks the grgit submodule isn't there.  What's weird is that this doesn't match up with `git status`.  I'm not sure how to use the command-line git to see what JGit is seeing.

**Comment by paleozogt** [comment]
I downloaded [jgit.sh](https://eclipse.org/jgit/download/) and tried it on this repo.  And it's also reporting that the submodule is modified:

```
> git status
On branch grgit_tests
Your branch is up-to-date with 'origin/grgit_tests'.
nothing to commit, working directory clean
>
> ~/Downloads/org.eclipse.jgit.pgm-3.7.1.201504261725-r.sh status
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
On branch grgit_tests
Changes not staged for commit:

	modified:   grgit
```

So this appears to be a JGit bug.  :(

**Comment by paleozogt** [comment]
After some more testing, I've realized this problem only happens on Mac OS X.  (I'm on 10.9.5.)

**Comment by paleozogt** [comment]
fyi, I've filed JGit [467631](https://bugs.eclipse.org/bugs/show_bug.cgi?id=467631)

**Comment by ajoberstar** [comment]
Thanks. If/when it gets fixed, I can bump the JGit version.

**Comment by ajoberstar** [comment]
Confirming that the JGit bug is still open as of 10/22/2017.

**Comment by vivin** [comment]
I've committed a fix for this on JGit [here](https://git.eclipse.org/r/126722).

**Comment by ajoberstar** [comment]
Just doing some cleanup. In theory, this should work now since the issue says it was fixed in 5.2 and grgit's on 5.9. Thanks @vivin for fixing. And feel free, anyone, to reopen if this is still occuring.
